Arc-wrap ClientConfig `RootCertStore` and remove expensiveness warnings

djc opened the pull request:
From #1403:

> We should add some nuance to this, explaining for instance that loading a bundle of roots from disk may be slow, so ClientConfigs/ServerConfigs that use that method should be reused to the extent possible. And perhaps going into even more detail, that the slowness specifically lives in [`ConfigBuilder::with_root_certificates`](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html#method.with_root_certificates) (and the steps that come before that, to load the RootCertStore).

I ended up just removing the remarks for now, after changing `with_root_certificates()` for the `ClientConfig` builder to take an `Arc<RootCertStore>` (the `ServerConfig` builder started taking an `Arc`-wrapped `RootCertStore` in #1368). IMO it would be best to document the performance implications where they occur, which in this case means `rustls_native_certs::load_native_certs()`. Its documentation currently reads:

> This function can be expensive: on some platforms it involves loading and parsing a ~300KB disk file. It’s therefore prudent to call this sparingly.

I also believe the use of `Arc` wrappers itself will help signal that (re)creating something of that type is performance-sensitive, so I feel there's not much value in further discussion directly on the config types (which, at worst, cost a few allocations for clones, many of which are actually refcount bumps from other `Arc`s).

Fixes #1403. Motivated by discussion in https://github.com/bluejekyll/trust-dns/issues/1990 (trust-dns-resolver currently stores a `Lazy<Arc<ClientConfig>>` wrapping webpki-roots -- which doesn't seem great).

daxpedda commented:
On that note, I'm aware that this is future talk here, but what is the plan for `rustls-platform-verifier`?
Will there be a way to use it without `dangerous_configuration`? Because if we decide to move [`ServerCertVerifier`](https://docs.rs/rustls/latest/rustls/client/trait.ServerCertVerifier.html) out of `dangerous_configuration` we might want to change that API a bit.

djc commented:
> On that note, I'm aware that this is future talk here, but what is the plan for `rustls-platform-verifier`? Will there be a way to use it without `dangerous_configuration`?

Conceptually I think there should be but it's not entirely obvious to me how that should work.

> Because if we decide to move [`ServerCertVerifier`](https://docs.rs/rustls/latest/rustls/client/trait.ServerCertVerifier.html) out of `dangerous_configuration` we might want to change that API a bit.

Like how?

daxpedda commented:
> > Because if we decide to move [`ServerCertVerifier`](https://docs.rs/rustls/latest/rustls/client/trait.ServerCertVerifier.html) out of `dangerous_configuration` we might want to change that API a bit.
> 
> Like how?

Probably expose [`WebPkiVerifier`](https://docs.rs/rustls/latest/rustls/client/struct.WebPkiVerifier.html) as well and change `with_root_certificates()` to take a `Arc<impl ServerCertVerifier>`.

djc commented:
> Probably expose [`WebPkiVerifier`](https://docs.rs/rustls/latest/rustls/client/struct.WebPkiVerifier.html) as well and change `with_root_certificates()` to take a `Arc<impl ServerCertVerifier>`.

It could make sense to drop `with_root_certificates()` in favor of `with_verifier(Arc<dyn ServerCertVerifier>)` (we already have the more elaborately named `with_custom_certificate_verifier()` today), and move the `RootCertStore` integration point to a `ServerCertVerifier` builder or constructor.

The open question is in that today, any foreign verifier wouldn't be able to impl `ServerCertVerifier` without also enabling `dangerous_configuration`. It might be nice if we found a way to do that. I suppose we could have a `PlatformVerifier` newtype wrapped around a rustls-platform-verifier type (to avoid leaking the rustls-platform-verifier version into the rustls public API) which could be enabled based on a Cargo feature. That might actually be an interesting pattern for webpki-roots and rustls-native-certs, too -- should make things easier for users.

jsha commented:
> Like how?

The nice thing about the builder API is that we can have multiple setters that affect the same underlying setting but use different types. For instance we could have:

```rust
fn with_platform_verifier(v: rustls_platform_verifier::ClientConfig)
```

This could be exposed in non-`dangerous_configuration` builds. Or, as you point out, make a newtype for it.

Similarly, for this change we could expose `with_arc_root_certificates`, deprecating `with_root_certificates`. I know we have a breaking change coming up anyway but this is a sort of low-stakes papercut that makes upgrades frustrating (and affects almost all downstream users). Providing both versions for a release cycle would make people's lives a bit easier.

For the expensiveness warning: Rather than delete it we should replace it with text that says the opposite ("making one of these is cheap"). People have read the existing warning and if it goes away they are likely to assume they simply can't find it or misremembered where it is, rather than that they no longer need to consider these objects expensive.

daxpedda commented:
But currently `rustls-platform-verifier` depends on `rustls`, so if we want to expose something like `with_platform_verifier()` or a newtype we would need to remove that first.

djc commented:
> > Like how?
> 
> The nice thing about the builder API is that we can have multiple setters that affect the same underlying setting but use different types. For instance we could have:
> 
> ```rust
> fn with_platform_verifier(v: rustls_platform_verifier::ClientConfig)
> ```
> 
> This could be exposed in non-`dangerous_configuration` builds. Or, as you point out, make a newtype for it.

Sure -- so far, though, we haven't "blessed" any verifier other than the webpki one by directly exposing it as part of the rustls API. This would change that, which feels like a larger conceptual change?

> Similarly, for this change we could expose `with_arc_root_certificates`, deprecating `with_root_certificates`. I know we have a breaking change coming up anyway but this is a sort of low-stakes papercut that makes upgrades frustrating (and affects almost all downstream users). Providing both versions for a release cycle would make people's lives a bit easier.

I do like the idea of providing `with_verifier()` and a public `WebPkiVerifier` with an `Arc<RootCertStore>` constructor on 0.21 so people can adapt.

@ctz thoughts?

> For the expensiveness warning: Rather than delete it we should replace it with text that says the opposite ("making one of these is cheap"). People have read the existing warning and if it goes away they are likely to assume they simply can't find it or misremembered where it is, rather than that they no longer need to consider these objects expensive.

Should we explicitly acknowledge that we used to say it was expensive?

ctz commented on rustls/src/client/builder.rs:
```diff
@@ -16,7 +16,7 @@ impl<C: CryptoProvider> ConfigBuilder<ClientConfig<C>, WantsVerifier<C>> {
     pub fn with_root_certificates(
         self,
-        root_store: webpki::RootCertStore,
+        root_store: Arc<webpki::RootCertStore>,
```
AIUI because there is a blanket implementation for that on all types this would make this much less invasive to users?

djc replied:
Yes, exactly.

djc replied:
Hmm, I guess that would also make it semver-compatible?

ctz commented:
> Ah yes, that's one reason this isn't quite trivial. I think the fix here would be that rustls-platform-verifier just offers an API (like rustls-webpki currently does) that mimics but doesn't implement the `ServerCertVerifier` API, and then the rustls crate would wrap the trait impl around it.

I'd agree with that approach (and seems like it would be marginally easier with the base types crate?). I've also found platform-verifier difficult to reuse/compose because of its dependency on rustls.

> > Similarly, for this change we could expose `with_arc_root_certificates`, deprecating `with_root_certificates`. I know we have a breaking change coming up anyway but this is a sort of low-stakes papercut that makes upgrades frustrating (and affects almost all downstream users). Providing both versions for a release cycle would make people's lives a bit easier.
> 
> I do like the idea of providing `with_verifier()` and a public `WebPkiVerifier` with an `Arc<RootCertStore>` constructor on 0.21 so people can adapt.
> 
> @ctz thoughts?

I don't think we're achieving anything by gating `WebPkiServerVerifier` behind `dangerous_configuration`.

djc commented:
> I don't think we're achieving anything by gating `WebPkiServerVerifier` behind `dangerous_configuration`.

But how do you feel about a change from `with_root_certificates(roots)` to `with_verifier(WebPkiServerVerifier::new(roots))`?

ctz commented:
> > I don't think we're achieving anything by gating `WebPkiServerVerifier` behind `dangerous_configuration`.
> 
> But how do you feel about a change from `with_root_certificates(roots)` to `with_verifier(WebPkiServerVerifier::new(roots))`?

I think that would be OK, it's certainly not breaking the bank on top of the quite significant amount of ceremony needed to get to that point (making a `RootCertStore`, getting the certs in there, etc.)

djc commented:
I've changed the `ServerCertVerifier` setup to take `impl Into<Arc<RootCertStore>>` so that the change is backwards compatible and replaced the "expensive" warning with a more precise explanation:

```
/// While creation of a `ServerConfig` was previously documented to be expensive, this has
/// recently been imprecise. The largest potential cost in building a `ServerConfig` is in
/// gathering trust roots from the operating system to add to the [`RootCertStore`] passed
/// to to a [`ClientCertVerifier`] builder (the rustls-native-certs is often used for this).
/// Other than that, the cost of instantiating is relatively cheap, involving a few allocations.
```

I've left out further changes to start taking an explicit verifier for the `ClientConfig`, since I think these will be more involve. If this is merged I'd like to backport it to the 0.21.x branch.

jsha commented on rustls/src/server/server_conn.rs:
```diff
@@ -185,8 +185,11 @@ impl<'a> ClientHello<'a> {
-/// Making one of these can be expensive, and should be
-/// once per process rather than once per connection.
+/// While creation of a `ServerConfig` was previously documented to be expensive, this has
+/// recently been imprecise. The largest potential cost in building a `ServerConfig` is in
```
```suggestion
/// recently become imprecise. The largest potential cost in building a `ServerConfig` is in
```

"recently been imprecise" is valid but reads weirdly to me because "recently" usually goes with a discrete moment in time rather than a continuous interval.

Coming to think of it, though - hasn't it always been the case that gathering the roots was the only expensive part?

How about:

> Making one of these is cheap, though one of the inputs may be expensive: gathering trust roots from the operating system (to call [`with_root_certificates`]) may take on the order of a few hundred milliseconds.

(Note that this doesn't explicitly say that it used to be expensive, which I think is okay because we're replacing the expensiveness text with other text that talks about expensiveness, rather than deleting it entirely).

djc replied:
Nice, I've used the language you've suggested.

jsha reviewed (changes requested):
Other than comment tweaks, this implementation plan and the backport look excellent to me.

djc commented on rustls/src/server/server_conn.rs:
```diff
@@ -185,8 +185,11 @@ impl<'a> ClientHello<'a> {
+/// While creation of a `ServerConfig` was previously documented to be expensive, this has
+/// recently been imprecise. The largest potential cost in building a [`ServerConfig`] is in
+/// gathering trust roots from the operating system to add to the [`RootCertStore`] passed
+/// to to a `ClientCertVerifier` builder (the rustls-native-certs is often used for this).
```
That's actually not so clear from the rustls-native-certs documentation. Maybe we should clear that up?
